llm/src/mosaic_gpt.py

```diff
@@ -30,40 +30,16 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
             batch_first=True,
             device=device,
         )
-
-        self.register_buffer(
-            'mask',
-            torch.empty((cfg.max_seq_len, cfg.max_seq_len), device=device))
-        self.mask_initialized = False
         self.mhsa.out_proj._is_residual = True  # type: ignore
 
-    def _fill_causal_attn_mask(self):
-        assert isinstance(self.mask, torch.Tensor)  # for type checking
-        torch.full(size=self.mask.shape,
-                   fill_value=float('-inf'),
-                   out=self.mask)
-        torch.triu(input=self.mask, diagonal=1, out=self.mask)
-
-    def forward(self, x, key_padding_mask):
-        # Two important disclaimers
-        # 1. Torch uses additive attention. If your attn_mask/key_padding mask is a float tensor, it will add the floats
-        #   directly to your attention matrix. If they are boolean masks, True will be converted to -inf before adding the
-        #   mask to your attentions. See https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention.forward
-        #   Basically True/-inf indicates tokens we do not want to attend to.
-        #
-        # 2. This is is the exact opposite behavior of Huggingface's tokenizers, which use the convention that True denotes tokens
-        #   we do want to attend to. See https://huggingface.co/docs/transformers/glossary#attention-mask
-        #
-        if not self.mask_initialized:
-            self._fill_causal_attn_mask()
-            self.mask_initialized = True
-
-        return self.mhsa(x,
-                         x,
-                         x,
-                         attn_mask=self.mask,
-                         key_padding_mask=~key_padding_mask,
-                         need_weights=True)
+    def forward(self, x, key_padding_mask, attn_mask=None):
```

Comment:
Does this mean Torch-native attention now works with torch 1.13? I had this listed as a bug but it looks like you may have fixed it? 👀

Reply:
not sure where the bug was / wasn't but [this test](https://github.com/vchiley/examples/blob/tritonFlashibiFaster/llm/tests/test_hf_v_mosaic_gpt.py#L18) checks the hf model against mosaic_gpt with torch attn and it passes.

Also the mask generation (including the comments you made about the mask generation) I moved to [where I generate the mask](https://github.com/vchiley/examples/blob/tritonFlashibiFaster/llm/src/mosaic_gpt.py#L250).